File: Backend/src/core/ai/providers/GeminiProvider.ts
// src/core/ai/providers/GeminiProvider.ts
import { GoogleGenerativeAI } from '@google/generative-ai';
import { IAiProvider, GenerateTextParams, GenerateImageParams } from '../IAiProvider';

export class GeminiProvider implements IAiProvider {
  private client: GoogleGenerativeAI;

  constructor(apiKey: string) {
    if (!apiKey) {
      throw new Error('Gemini API key is required.');
    }
    this.client = new GoogleGenerativeAI(apiKey);
  }

  async generateText(params: GenerateTextParams): Promise<string> {
    try {
      const model = this.client.getGenerativeModel({ model: params.model });
      const result = await model.generateContent(params.prompt);
      const response = await result.response;
      return response.text();
    } catch (error) {
      console.error('Error generating text with Gemini:', error);
      throw new Error('Failed to generate text using Gemini.');
    }
  }

  async generateImage(params: GenerateImageParams): Promise<string | null> {
    // Nota: La API de Gemini para generación de imágenes (Imagen) puede tener un SDK o requerimientos diferentes.
    // Esta es una implementación conceptual. Asumimos que el SDK puede manejarlo.
    // En una implementación real, se usaría el SDK específico para 'Imagen'.
    console.warn('Gemini image generation is conceptually implemented. Requires Imagen API setup.');

    // Simulación de llamada a la API de Imagen.
    // La implementación real requeriría una librería o endpoint REST diferente.
    try {
        const model = this.client.getGenerativeModel({ model: "imagen-4.0-generate-001" }); // Usar el modelo correcto

        // La API real podría variar. Esto es una aproximación.
        const result = await model.generateContent(`Generate an image based on this prompt: ${params.prompt}`);
        const response = await result.response;

        // El procesamiento de la respuesta para obtener la URL de la imagen dependería del formato de la API.
        // Aquí asumimos que devuelve un texto con la URL o datos base64.
        const textResponse = response.text();
        if (textResponse.startsWith('data:image')) {
            return textResponse;
        }
        return null;

    } catch (error) {
        console.error('Error generating image with Gemini (Imagen):', error);
        throw new Error('Failed to generate image using Gemini.');
    }
  }
}

File: Backend/api/services/ai_manager/providers/gemini_provider.py
# api/services/ai_manager/providers/gemini_provider.py
import google.generativeai as genai
from ..ai_base_provider import AIBaseProvider
from typing import Optional, Literal

class GeminiProvider(AIBaseProvider):
    """
    Proveedor de IA para Google Gemini.
    """

    def __init__(self, api_key: str):
        if not api_key:
            raise ValueError("Google Gemini API key is required.")
        genai.configure(api_key=api_key)

    def generate_text(self, prompt: str, model: str, **kwargs) -> str:
        try:
            model_instance = genai.GenerativeModel(model)
            response = model_instance.generate_content(prompt)
            return response.text
        except Exception as e:
            print(f"Error generating text with Gemini: {e}")
            raise RuntimeError("Failed to generate text using Gemini.") from e

    def generate_image(
        self,
        prompt: str,
        model: str,
        size: Optional[Literal['256x256', '512x512', '1024x1024']] = None,
        **kwargs
    ) -> Optional[str]:
        # Nota: Esta es una implementación conceptual para la generación de imágenes.
        # El SDK puede tener un método o modelo específico para 'Imagen'.
        print("Conceptual implementation for Gemini image generation.")
        try:
            # La generación de imágenes podría requerir un modelo específico como 'imagen'
            # y un método diferente. Asumimos que el SDK puede manejarlo así para este ejemplo.
            model_instance = genai.GenerativeModel(model)
            # El prompt podría necesitar un formato específico.
            response = model_instance.generate_content(f"Generate an image of: {prompt}")

            # El procesamiento de la respuesta para obtener la URL o base64 dependerá de la API.
            # Aquí asumimos que podría estar en el texto de la respuesta.
            if response.text.startswith("data:image"):
                return response.text
            return None
        except Exception as e:
            print(f"Error generating image with Gemini: {e}")
            raise RuntimeError("Failed to generate image using Gemini.") from e
